### DISCUSSION TITLE
How do I save the generated function of a derivative for later use?

### DISCUSSION THREAD

**Discussion by ejmeitz** [brianguenter/FastDifferentiation.jl#57] [Q&A] [answered]
Hello,

I have several derivative calculations that I would like to save and be able to call inside of a package without the need to re-calculate them every time I load the package. I'm sure I could rely on the pre-compile cache but I was wondering if there's a way to take the generated function and make it callable after I generate them?

This is maybe more of a Julia question rather than FastDifferentiation but thought I'd ask here first as it is your function I'd like to save.

Thanks!

**Comment by brianguenter** [comment]
Try JLD2.jl. I haven't used it myself but it seems to work well.

**Reply by ejmeitz** [reply]
Using just the defaults for JLD2 does not quite seem to work. I did the following:

```julia
using FastDifferentiation
@variables x y
f1 = cos(x) *y
f2 = sin(y) * x
symb = jacobian([f1,f2],[x,y])
jac_exe = make_function(symb, [x,y])
jldsave(<PATH>, jac = jac_exe)
```
This code runs fine, but after restarting and loading the function I have issues. Upon loading I get a warning that to my eye does not seem like a problem but I also am not familiar with generated functions.
```julia
using JLD2
jac_exe = load(<PATH>, "jac")
jac_exe([1.0,2.0])
```
> ┌ Warning: type RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:input_variables,),FastDifferentiation.#_RGF_ModTag,FastDifferentiation.#_RGF_ModTag,(0x691da5ea, 0x633a6af5, 0xa9eb32b9, 0x48835460, 0xf5e60f6d),Expr} does not exist in workspace; reconstructing
> └ @ JLD2 C:\Users\ejmei\.julia\packages\JLD2\cHcDY\src\data\reconstructing_datatypes.jl:605
> 

The error comes when you try to call the function:
`jac_exe([1.0,2.0]`

> ERROR: MethodError: objects of type JLD2.ReconstructedMutable{Symbol("RuntimeGeneratedFunction{(:input_variables,),#_RGF_ModTag,#_RGF_ModTag,(0x691da5ea, 0x633a6af5, 0xa9eb32b9, 0x48835460, 0xf5e60f6d),Expr}"), (:body,), …} are not callable

**Comment by brianguenter** [comment]
Yeah, I looked more closely at the JLD2 docs. Not sure it can serialize functions. I've never tried to do this so you're beyond my expertise level. Discourse would be a good place to ask this question. 

An alternative workaround might be to use the FastDifferentiation function `make_Expr` to make an `Expr` then write that `Expr` to disk, potentially with JLD2, then read it from disk and `eval` it to generate the function. 

Or maybe you could serialize the `Expr` using `Serialization.serialize` and read it back in with `Serialization.deserialize` then try to `eval` it. 

I've never tried either of these. Might work.

Here's how to create an `Expr` from your function:
```julia
julia> @variables x
x

julia> f = x^2
(x ^ 2)

julia> make_Expr([f],[x],true,true)
:((result, input_variables)->begin
          #= C:\Users\seatt\.julia\packages\FastDifferentiation\mftkg\src\CodeGeneration.jl:188 =#
          #= C:\Users\seatt\.julia\packages\FastDifferentiation\mftkg\src\CodeGeneration.jl:188 =# @inbounds begin
                  #= C:\Users\seatt\.julia\packages\FastDifferentiation\mftkg\src\CodeGeneration.jl:189 =#
                  begin
                      result_element_type = eltype(input_variables)
                      var"##226" = input_variables[1] ^ 2
                      result[1] = var"##226"
                      return nothing
                  end
              end
      end)
```

**Reply by ejmeitz** [reply]
Well I can save and load the expressions fine, but how would you make this expression executable? I tried just doing like `jac_exe = eval(<result of Make_Expr>)` and this is executable but it errors (even if I don't save to file and load it back).

**Comment by brianguenter** [comment]
Do you have a MWE of the eval of make_expr not working? I can take a look
tomorrow.

**Reply by ejmeitz** [reply]
Basically copied from the examples, except instead of using `make_function` I did `make_Expr` and then evaluated it to make the function. The error is on the last line:
`MethodError: no method matching (::var"#3#4")(::Vector{Float64})`
```julia
using FastDifferentiation
@variables x y
f1 = cos(x) * y
f2 = sin(y) * x
symb = jacobian([f1,f2], [x,y]);
expression = make_Expr([f1,f2],[x,y], true, true)
evaled = eval(expression)
evaled([1.0,2.0])
```

Is there a purpose to the commented out lines when you generate a function or expression? If these paths matter it might not be possible (at least easily) to re-use these functions in a package.

> #= C:\Users\ejmei\.julia\packages\FastDifferentiation\YGbNU\src\CodeGeneration.jl:200 =#
> #= C:\Users\ejmei\.julia\packages\FastDifferentiation\YGbNU\src\CodeGeneration.jl:200 =# @inbounds begin
> #= C:\Users\ejmei\.julia\packages\FastDifferentiation\YGbNU\src\CodeGeneration.jl:201 =#

Also, from the `Symbolics.jl` library there is a `build_function` with `JuliaTarget` which looks like it does something similar to what I hoping for (have not tested this, but I know Symbolics could calculate the derivatives I need).
https://symbolics.juliasymbolics.org/stable/manual/build_function/#build_function

**Comment by brianguenter** [comment]
You created the expression with the evaluate in place argument set to true.
If you look at the function signature of the generated function you will
see that it expects two arguments: the first argument is an array which
will hold the result and the second is an array of input values.


You then called the generated function with a single argument. There was no
such function so the Julia run time reported an error.

Make_expr and make_function work the same way. Check the documentation for
details and examples. And if you think the documentation could be improved
submit a PR.

**Reply by ejmeitz** [reply]
Thanks! This seems to work. I was able to save the raw expression to a JLD2 file which I then moved to a different computer. I loaded the expression and evaluated it and it gave the expected output. 

Are those commented out paths just a history of what generated the code?

**Comment by brianguenter** [comment]
Great to hear it's working.

I believe those are created by RuntimeCode Generation.jl to show what
source line gave rise to the generated code. FastDiferentiation.jl doesn't
generate those lines.

**Comment by ejmeitz** [comment, marked as answer]
Just so there's something to mark as an answer. The following code works:
```julia
using FastDifferentiation
using JLD2
@variables x y
f1 = cos(x) * y
f2 = sin(y) * x
symb = jacobian([f1,f2], [x,y]);
expression = make_Expr([f1,f2],[x,y], false, true)
jldsave(<path>, expr = expression)
.......
expression = load(<path>, "expr")
evaled = eval(expression)
evaled([1.0,2.0])
```

**Comment by brianguenter** [comment]
If you have time could you do a PR to the documentation under the heading
"writing derivatives to disk"?

**Reply by ejmeitz** [reply]
yep for sure, I'm gonna look at what symbolics.jl does as well they have similar functionality

**Reply by ejmeitz** [reply]
I looked at `Symbolics.jl` they have [this ](https://docs.sciml.ai/Symbolics/stable/manual/io/) page. You can literally just convert the `Symbol` from `make_Expr` to a `String` (and wrap with function tag) and save it to a file which you can then import to a package. Seems obvious now that I see it but a rather elegant solution I think.

There's a couple other ways to do this, I'll try and implement a couple and explain them when I make the PR.
